subprojects/base-services/src/main/java/org/gradle/internal/time/MonotonicClock.java

```diff
@@ -70,28 +66,32 @@
     @VisibleForTesting
     MonotonicClock(TimeSource timeSource, long syncIntervalMillis) {
         this.timeSource = timeSource;
-        this.syncMillis = this.max = timeSource.currentTimeMillis();
-        this.syncNanos = timeSource.nanoTime();
         this.syncIntervalMillis = syncIntervalMillis;
+        this.syncNanosRef = new AtomicLong(timeSource.nanoTime());
+        this.max.set(this.syncMillis = timeSource.currentTimeMillis());
     }
 
     public long getCurrentTime() {
-        lock.lock();
-        try {
-            long nowNanos = timeSource.nanoTime();
-            long sinceSyncNanos = nowNanos - syncNanos;
-            long sinceSyncMillis = TimeUnit.NANOSECONDS.toMillis(sinceSyncNanos);
-            long currentTime = syncMillis + sinceSyncMillis;
+        long nowNanos = timeSource.nanoTime();
+        long syncNanos = syncNanosRef.get();
+        long sinceSyncNanos = nowNanos - syncNanos;
+        long sinceSyncMillis = TimeUnit.NANOSECONDS.toMillis(sinceSyncNanos);
 
-            if (sinceSyncMillis >= syncIntervalMillis) {
-                syncNanos = nowNanos;
-                return syncMillis = max = Math.max(timeSource.currentTimeMillis(), max);
-            } else {
-                return max = Math.max(currentTime, max);
-            }
-        } finally {
-            lock.unlock();
+        if (sinceSyncMillis >= syncIntervalMillis && syncNanosRef.compareAndSet(syncNanos, nowNanos)) {
+            return syncMillis = issue(timeSource.currentTimeMillis());
```

Comment:
⭕️ There is no coordination between syncNanosRef, max and syncMillis. I believe this is only safe because we have a long resync interval that makes sure two requests don't run into each other here. Might be worth documenting the strategy as a method-level comment.

Reply:
Yes, that's the case.